File: scripts/Colorize.hs
{-
Installation:

sudo apt-get install ghc
cabal update
cabal install ansi-terminal

runhaskell Colorize.hs
- or -
ghc --make Colorize.hs -o Colorize

Then just create a script to pipe your junk through Colorize.
make $* 2>&1 | Colorize
exit ${PIPESTATUS[0]}
-}

module Main where

import System.Console.ANSI
import Data.List
import Text.Regex
import Text.Regex.Posix
import System.Posix.Directory
import System.IO.Unsafe
import System.IO

--- Config ---
markup = [ ("error:",           colorPutStrLn BoldIntensity Vivid Cyan)
         , ("!!!FAILURES!!!",   colorPutStrLn NormalIntensity Vivid Cyan)
         , ("undefined",        colorPutStrLn NormalIntensity Vivid Cyan)
         , ("assertion",        colorPutStrLn NormalIntensity Vivid Cyan)
         , ("within",           colorPutStrLn NormalIntensity Dull Cyan)
         , ("note:",            colorPutStrLn NormalIntensity Vivid White)
         , ("warning:",         colorPutStrLn NormalIntensity Vivid Yellow)
         , ("instantiated",     colorPutStrLn NormalIntensity Vivid Green)
         , ("from",             colorPutStrLn NormalIntensity Vivid Green)
         , ("In",               colorPutStrLn NormalIntensity Dull Green)
         , ("g++",              colorPutStrLn NormalIntensity Dull Yellow)
         , ("ar",               colorPutStrLn NormalIntensity Dull Yellow)
         , ("perl",             colorPutStrLn NormalIntensity Dull Yellow)
         , ("mock.py",          colorPutStrLn NormalIntensity Dull Yellow)
         , ("cpplint",          colorPutStrLn NormalIntensity Dull Yellow)
         , ("make:",            colorPutStrLn BoldIntensity Dull Green)
         ]

ss = "std::basic_string<char, std::char_traits<char>, std::allocator<char> >"
substs = [ (ss, "string")
         , ("python cpplint.py", "cpplint")
         ]

prefixsToStrip = [cwd ++ "/src/", cwd ++ "/"]
  where cwd = unsafePerformIO getWorkingDirectory


cleanup = id
        |. elideCompiles ["g++", "ar"]
        |. killFollowing "perl"
        |. killFollowing "python cpplint.py"
        |. stripPaths
        |. applySubsts
        |. markLine

------

-- | Just to make things look like pipes instead of composition.
(|.) = flip (.)
infixl 9 |.

-- | Kill all prefixesToStrip from a line
stripPaths :: String -> String
stripPaths line = foldl stripIfPrefixed line prefixsToStrip

-- | Kill prefix from line.
stripIfPrefixed :: String -> String -> String
stripIfPrefixed line prefix =
  case stripPrefix prefix line of
    Just s -> s
    Nothing -> line

-- | See elideCompile.  Elides compiles for every lines starting with a word
--   in a list.
elideCompiles :: [String] -> String -> String
elideCompiles = flip $ foldl elideCompile

-- | If the first word of line is word then delete all but the first and last
--   word in the line.
elideCompile :: String -> String -> String
elideCompile line word =
  if not (null w) && head w == word
    then unwords [head w, stripPaths $ last w]
    else line
  where w = words line

-- | Replace all words in the line beyond the first with "..." if the first
--   word matches the given word.
killFollowing word line =
  if line =~ ("^" ++ word)
     then unwords [word, "..."]
     else line

-- | Apply pairs from substs on the line.  The first string in each pair is a
--   regular expression pattern which is replaced with the second string from
--   the pair.  The pairs are applied in order.
applySubsts :: String -> String
applySubsts line = foldl sub line substs
    where sub l (pat, repl) = subRegex (mkRegex pat) l repl

-- | Apply pairs from markup on the line.  For a string from the pair that
--   matches the corresponding action from the pair is performed on the line.
--   Pairs are applied in order; the lastmost match is the action that gets
--   performed.
markLine :: String -> IO ()
markLine l = foldl combine putStrLn markup $ l
    where combine action (word, action') =
            if word `elem` words l
              then action'
              else action

-- | Print line with consoleIntensity, colorIntensity and color.
colorPutStrLn consoleIntensity colorIntensity color line = do
  setSGR [SetConsoleIntensity consoleIntensity,
          SetColor Foreground colorIntensity color]
  putStrLn line
  setSGR [Reset]
  hFlush stdout

main = do
    contents <- getContents
    sequence_ $ map cleanup $ lines contents

File: scripts/colorize.py
#!/usr/bin/env python
# Copyright (c) 2012 Stanford University
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Installation:
#
# Install termcolor available here or through your distro:
# http://pypi.python.org/pypi/termcolor
# 
# Then just create a script to pipe your junk through colorize.
# make $* 2>&1 | python colorize.py
# exit ${PIPESTATUS[0]}

import termcolor
import re
import fileinput
import os
import sys

# Most of the 'configuration'.
# Default substitutions, colors, etc.

def color(*args, **kwargs):
    """Return a function that takes a string and colors it."""
    return lambda s: termcolor.colored(s, *args, **kwargs)

markup = [ ('g++',              color('yellow'))
         , ('ccache',           color('yellow'))
         , ('ar',               color('yellow'))
         , ('perl',             color('yellow'))
         , ('mock.py',          color('yellow'))
         , ('cpplint',          color('yellow'))
         , ('make:',            color('green'))
         , ('error:',           color('red', attrs=['bold']))
         , ('Error',            color('red', attrs=['bold']))
         , ('!!!FAILURES!!!',   color('cyan', attrs=['bold']))
         , ('undefined',        color('cyan', attrs=['bold']))
         , ('assertion',        color('cyan', attrs=['bold']))
         , ('within',           color('cyan'))
         , ('note:',            color('white'))
         , ('warning:',         color('yellow'))
         , ('instantiated',     color('green'))
         , ('from',             color('green'))
         , ('In',               color('green'))
         , ('',                 color())
         ]

ss = 'std::basic_string<char, std::char_traits<char>, std::allocator<char> >'
substs = [ (ss, 'string')
         , ('boost::intrusive_ptr<RAMCloud::Transport::Session>',
            'Transport::SessionRef')
         , (r'^perl .*$', 'perl ...')
         , (r'^python cpplint\.py.*$', 'cpplint')
         ]

prefixesToStrip = [ os.getcwd() + '/src/'
                  , os.getcwd() + '/'
                  ]
prefixesToStrip.sort(key=len, reverse=True) # sort prefixes by length desc

def cleanup(line):
    line = line.strip()
    line = elideCompiles(['ccache', 'g++', 'ar'], line)
    line = stripPaths(line)
    line = applySubsts(line)
    markLine(line)

# Many utility functions for slicing and dicing lines.

def stripPaths(line):
    """Kill all prefixesToStrip from a line"""
    def stripWord(word):
        for prefix in prefixesToStrip:
            if word.startswith(prefix):
                return word[len(prefix):]
        return word
    return ' '.join([stripWord(word) for word in line.split()])

def elideCompiles(words, line):
    """If the first word of line is in words then delete all but the first
       and last word in the line."""
    for word in words:
        if line.startswith(word):
            lwords = line.split()
            return '%s %s' % (lwords[0], lwords[-1])
    return line

def applySubsts(line):
    """Apply pairs from substs on the line.  The first string in each pair "
    is a regular expression pattern which is replaced with the second "
    string from the pair.  The pairs are applied in order."""
    for left, right in substs:
       line = re.sub(left, right, line)
    return line

def markLine(line):
    """Apply pairs from markup on the line.  For a string from the pair
       that matches the corresponding markup from the pair is performed
       on the line.  Pairs are applied in order; the lastmost match is
       the action that gets performed."""
    for word, action in markup:
        if word in line:
            print(action(line))
            sys.stdout.flush()
            return

def main():
   while True:
       line = sys.stdin.readline()
       if line == "":
           break
       cleanup(line)

if __name__ == '__main__': main()
